common/migration/read.go

```diff
@@ -0,0 +1,163 @@
+
+	// run the shadow request in a goroutine, wait for the results and then gather them up
+	go func() {
+		defer func() {
+			log.CapturePanic(recover(), c.log, nil)
+			c.scope.IncCounter(metrics.ShadowerBackgroundPanicRecoverCount)
+		}()
+		defer wg.Done()
+
+		backgroundRes, backgroundErr := c.callBackgroundAndCompare(ctx, background)
+		results[1].res = backgroundRes
+		results[1].err = backgroundErr
+	}()
+
+	// Fire off a watcher which is waiting for both sets of results to come through, and then once done so
+	// it'll hand them to the comparison function for analysis
+	go func() {
+		if !c.shouldCompare(constraints) {
+			return
+		}
+		defer func() {
+			log.CapturePanic(recover(), c.log, nil)
+			c.scope.IncCounter(metrics.ShadowerBackgroundPanicRecoverCount)
+		}()
+
+		wg.Wait()
+		c.comparisonFn(c.log, c.scope, results[0].res, results[0].err, results[1].res, results[1].err)
+	}()
+
+	activeRes, activeErr := active(ctx)
+	results[0].res = activeRes
+	results[0].err = activeErr
+	wg.Done()
+	return activeRes, activeErr
+}
+
+// Fires off a request on a background goroutine and waits for it to complete OR when the hard deadline
+// in backgroundTimeout expires. The op() code is expected to be faulty, to possibly panic, take too long
+// or otherwise misbehave and so the hard deadline is intended as a last-ditch safety measure if the context
+// is not being obeyed
+func (c *readerImpl[T]) callBackgroundAndCompare(parentCtx context.Context, op func(ctx context.Context) (*T, error)) (*T, error) {
+
+	newRequestCtx, cancel := copyParentContextWhileBreakingParentCancellation(parentCtx)
+	defer cancel()
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), c.backgroundTimeout)
+	defer cancel()
+
+	completion := make(chan resultPair[*T])
+	go func() {
+		defer func() {
+			log.CapturePanic(recover(), c.log, nil)
+			c.scope.IncCounter(metrics.ShadowerBackgroundPanicRecoverCount)
+		}()
+		defer close(completion)
+
+		res, err := op(newRequestCtx)
+		completion <- resultPair[*T]{
+			res: res,
+			err: err,
+		}
+	}()
+
+	// now return whatever finishes first
+	select {
+	case <-timeoutCtx.Done():
+		c.scope.IncCounter(metrics.ShadowerBackgroundRequestHardCutCount)
+		return nil, fmt.Errorf("shadower cut off a request taking too long: %w", timeoutCtx.Err())
+	case resPair := <-completion:
+		return resPair.res, resPair.err
+	}
+}
+
+// the intention here is to allow this child context to continue running until completion
+// because in shadowing, you're necessarily not going to want to block on a slow/maybe not working
+// new codepath, but we do need to wait for results for both (to compare). So when the active flow
+// is finishing first, we want to not have it's cancellation kill off the background context.
+func copyParentContextWhileBreakingParentCancellation(ctx context.Context) (c context.Context, cancelFunc func()) {
+	parentDeadline, ok := ctx.Deadline()
+	if !ok {
+		return ctx, func() {}
```

Comment:
hmm.  so if there is no deadline, we want to allow this to cancel it:
```go
go func() {
  time.Sleep(...)
  cancel()
}()
```
but if there is, we don't?

kinda feels... unintended.  like no-deadline means it should be `context.WithoutCancel(ctx)` so cancel consistently does not propagate.

---

that said, I generally think not propagating cancel is too risky to be worth it.  e.g. how does this handle service shutdown?  this'd probably need a service-lifecycle-aware background-comparison-monitor to deal with that, and we _should_ deal with it.

Reply:
oh, the If statement thing? Yeah, I admit I didn't think about that thoroughly, but yeah,  Ithink you're right. 

I'm not sure I understand your second point? The problem I'm trying to solve is handling the case where the old (fast, safe) path has completely returned and called cancel and the new/slow/has problems path gets cancelled by this and we don't get a chance to understand what went wrong or compare results.

Reply:
second part was mostly about "if we want to block cancels, we probably need an additional lifecycle manager" because otherwise we won't know when (or if) it completes, and that's probably relevant during shutdown.